### ISSUE TITLE
using hipCL for OpenCL interoperability

### ISSUE THREAD

**Issue by DieGoldeneEnte** [illuhad/hipSYCL#198]
I had a quick look at [hipCL](https://github.com/cpc/hipcl). Surprisingly of all the functions mentioned as not fully implemented, I only found 2 (using grep):
* `hipStreamCreateWithFlags` in `queue.cpp` and `hip_queue.cpp` to create non-blocking streams
* `hipGetDeviceProperties` in `device.hpp`

From my (limited) understanding, the `hipStreamCreateWithFlags` should be no problem. As the README states:
> The flags change how the runtime waits for results (yield thread to OS or busy waiting / spinning)

The `hipGetDeviceProperties` function returns values, but some (not currently used in hipSYCL) are made up.

In my opinion hipSYCL should work with hipCL, but I didn't test it yet. If I got my environment setup and some time (likely in mid February) I might try compiling a program using hipCL.

Properties made up by hipCL:
* regsPerBlock
* warpSize
* maxGridSize
* memoryClockRate
* memoryBusWidth
* major
* minor
* maxThreadsPerMultiProcessor
* computeMode
* arch
* arch.hasGlobalInt32Atomics
* arch.hasSharedInt32Atomics
* arch.hasGlobalInt64Atomics
* arch.hasSharedInt64Atomics
* clockInstructionRate
* concurrentKernels
* pciDomainID
* pciBusID
* pciDeviceID
* isMultiGpuBoard
* canMapHostMemory
* gcnArch
* integrated

**Comment by illuhad** [comment]
Thank you for investigating! :) `hipStreamCreateWithFlags` is very important to get rid of CUDA default stream semantics (there's a flag that tells it to be non-blocking, otherwise all operations implicitly synchronize with the default stream for ... legacy CUDA reasons). 
This is important in order to have multiple streams execute concurrently. If this isn't supported in HipCL, it may indicate some significant performance impact for more complex applications as things like overlapping compute and data transfers will be impossible.

**Comment by DieGoldeneEnte** [comment]
For testing it should still work, since the queue is still created and the flags are passed to the ClQueue. Maybe some flags will work, but this requires further testing.

From the implementation:
> // TODO priority & flags require an OpenCL extensions
so this might be fixed in the future.

**Comment by illuhad** [comment]
If they just create an OpenCL command queue per stream it would mean they probably ditched the ridiculous CUDA/HIP default stream semantics. This would give us the right behavior, even if the flag is not supported.

**Comment by fodinabor** [comment]
Yeah.. just looked at the code and it looks like they do it _the right way_.
In `hipStreamCreateWithFlags` a new stream is created:
https://github.com/cpc/hipcl/blob/35cb616269cd81cd3652b19614e41e6f8d0c7219/lib/hipcl.cc#L438
which actually results in a new OpenCL ` cl::CommandQueue`:
https://github.com/cpc/hipcl/blob/43afc2ed34715e8a09967f71a79a806d5f295e9c/lib/backend.cc#L762

and I haven't found any unnecessary synchronization elsewhere..

**Comment by illuhad** [comment]
@fodinabor excellent news, thank you!
